From 4404d74efd6e5c4f088695476ee2794bd5c270b2 Mon Sep 17 00:00:00 2001
From: Grazvydas Ignotas <notasas@gmail.com>
Date: Fri, 8 Apr 2016 19:21:36 +0300
Subject: [PATCH 1/2] don't do SELinux check on Android

Some Android versions use SELinux, but it doesn't seem to complain about
write+exec mappings.
Also the check depends on getline(), which older versions of NDK lack.
---
 src/closures.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/closures.c b/src/closures.c
index 3dec0e3..d24d9b0 100644
--- a/src/closures.c
+++ b/src/closures.c
@@ -55,7 +55,7 @@
 #endif
 
 #if FFI_MMAP_EXEC_WRIT && !defined FFI_MMAP_EXEC_SELINUX
-# ifdef __linux__
+# if defined(__linux__) && !defined(__ANDROID__)
 /* When defined to 1 check for SELinux and if SELinux is active,
    don't attempt PROT_EXEC|PROT_WRITE mapping at all, as that
    might cause audit messages.  */

From 0ca8c959f0d219c49ec9d21b9e935f944a698320 Mon Sep 17 00:00:00 2001
From: Grazvydas Ignotas <notasas@gmail.com>
Date: Fri, 8 Apr 2016 19:39:35 +0300
Subject: [PATCH 2/2] use explicit write+exec mapping on Android

On Linux, by default libffi relies on malloc() memory to have
write+execute permissions, which is usually comes as a side effect of
making the stack executable (either by execstack linker option or
by compiling and linking some .S files; libffi linked programs get
linked to such objects).

On Android, at least on some versions of NDK for ARM, it is not enough
to just link some objects compiled from .S (or libffi) to enable
execstack and to get executable heap, so libffi just crashes in such
configuration. It can be argued it's not even desirable to make the
whole heap executable because of security implications, so enable
FFI_MMAP_EXEC_WRIT to make only needed memory executable instead.
---
 configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index da7c432..b340087 100644
--- a/configure.ac
+++ b/configure.ac
@@ -197,7 +197,7 @@ case "$target" in
                  [Cannot use PROT_EXEC on this target, so, we revert to
                    alternative means])
      ;;
-     *-apple-darwin* | *-*-freebsd* | *-*-kfreebsd* | *-*-openbsd* | *-pc-solaris*)
+     *-apple-darwin* | *-*-freebsd* | *-*-kfreebsd* | *-*-openbsd* | *-pc-solaris* | *-linux-android*)
        AC_DEFINE(FFI_MMAP_EXEC_WRIT, 1,
                  [Cannot use malloc on this target, so, we revert to
                    alternative means])
